### PR TITLE
Raise error when non-existent enum used in query

### PR DESCRIPTION
**Pull request by ak15**
This behaviour is in
rails/activerecord/lib/active_record/enum.rb #serialize(value) line no 143
if value is not present in mapping we are sending the value back ,
which in mysql returns unrelated record

**Going Forward we will raise error if enum value not present.**

### Summary
A non-exist enum string , gets unrelated record in MySQL.
This pull request solves the issue by setting a value nil, if it is not present among the enum.
This solves the following issue mentioned with example below
https://github.com/rails/rails/issues/38687
### Other Information
```
class User < ActiveRecord::Base
  enum status: { active: 0, non_active: 1 }
 end
 User.where(status: :non_existing_status)
 => ArgumentError ('non_existing_status' is not a valid status)
```

### CODE REVIEW

**Comment by ak15** [comment]
Can someone from Rails team please review and give feedback  on  this.
Its my first pull request in rails, will be happy and blessed to contribute more.

File: activerecord/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,19 @@
...
+  
+  When passing non existed enum in query, it return unrelated record in MySql.
+  Fixes #38687
+  
+    ```ruby
+    class User < ActiveRecord::Base
+      enum status: { active: 0, non_active: 1 }
+    end
+    User.where(status: :non_existing_status) 
+    # Above returns unrelated record for existing code
```
Comment (jeremy) [inline activerecord/CHANGELOG.md]: Seems this should raise an exception

File: activerecord/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,19 @@
...
+    User.where(status: User.statuses[:non_existing_status]) 
+    # Above returns record with nil status
```
Comment (jeremy) [inline activerecord/CHANGELOG.md]: and this is `User.where(status: nil)`, so should return records with nil status

File: activerecord/lib/active_record/enum.rb [hunk 3]
```diff
@@ -140,6 +140,7 @@ def deserialize(value)
       end
 
       def serialize(value)
+        return unless mapping.has_key?(value) || mapping.has_value?(value)
         mapping.fetch(value, value)
```
Comment (jeremy) [inline activerecord/lib/active_record/enum.rb]: Looks like we're attempting to implement `#cast` in serialize here? Do we want to `assert_valid_value(value)`?

File: activerecord/test/cases/enum_test.rb [hunk 4]
```diff
@@ -59,6 +59,7 @@ class EnumTest < ActiveRecord::TestCase
     assert_not_equal @book, Book.where(status: [written]).first
     assert_not_equal @book, Book.where("status <> ?", published).first
     assert_equal @book, Book.where("status <> ?", written).first
+    assert Book.where(status: Book.statuses[:not_defined]).blank?
```
Comment (jeremy) [inline activerecord/test/cases/enum_test.rb]: `assert_empty`

**Comment by ak15** [comment]
@jeremy 
Thanks for you valuable feedback, done with all changes mentioned.
 Squashed all commits into 1 commit .

File: activerecord/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,20 @@
...
+    class User < ActiveRecord::Base
+      enum status: { active: 0, non_active: 1 }
+    end
+    User.where(status: :non_existing_status) 
+    # Above returns unrelated record.
+    # Going forward will raise an exception.
+    User.where(status: nil) 
+    # Above returns record with nil status.
+    ```
+  *Atul Kanswal *
```
Comment (eileencodes) [inline activerecord/CHANGELOG.md]: This changelog makes it sound like we're introducing a bug when you're actually fixing a bug. Can you reword it to the following:

```
* Raise error when non-existent enum used in query

This change will raise an error when a non-existent enum is passed to a query. Previously
with MySQL this would return an unrelated record. Fixes #38687.
```

And update the code example to:

 ```ruby
    class User < ActiveRecord::Base
      enum status: { active: 0, non_active: 1 }
    end
    User.where(status: :non_existing_status) 
    => ArgumentError ('not_defined' is not a valid status)
  ```

Comment (ak15) [reply activerecord/CHANGELOG.md]: 🙏
Sure.
Special thanks for your talk [Breaking Down the Barrier: Demystifying Contributing to Rails](https://www.youtube.com/watch?v=7zoD6NZy6vY&t=3s).

**Comment by eileencodes** [comment]
Hey @ak15, I had to revert this. Once we tried to upgrade GitHub this week we realized there are cases where apps might be accepting unknown enums and that suddenly raising for this could cause a bad user experience. We need to deprecate this behavior first instead of raising an error. Let me know if you're unsure how to do that. We'll post a script or two to demonstrate how this could negatively affect apps that aren't expecting this change.

**Comment by ak15** [comment]
> Hey @ak15, I had to revert this. Once we tried to upgrade GitHub this week we realized there are cases where apps might be accepting unknown enums and that suddenly raising for this could cause a bad user experience. We need to deprecate this behavior first instead of raising an error. Let me know if you're unsure how to do that. We'll post a script or two to demonstrate how this could negatively affect apps that aren't expecting this change.
> 
> Revert commit [f3dfed7](https://github.com/rails/rails/commit/f3dfed7d0634bb7c07d761c62c11065d82488941)

Hi @eileencodes ,
Would be great help if you post a script to demonstrate how this negatively affect apps.

Since currently passing non-existent enum in query , gives unrelated record . This might cause unexpected errors. Please correct me if wrong.

Yes this creates a bad user experience, hence  created pull request with deprecation warning
https://github.com/rails/rails/pull/38826#issue-394369073
